designs/v1-api.md

```diff
@@ -0,0 +1,535 @@
+        kind: EC2NodeClass
+        name: default
+      taints:
+        - key: example.com/special-taint
+          effect: NoSchedule
+      startupTaints:
+        - key: example.com/another-taint
+          effect: NoSchedule
+      requirements:
+        - key: "karpenter.k8s.aws/instance-category"
+          operator: In
+          values: ["c", "m", "r"]
+          minValues: 2 # Alpha field, added for spot best practices support
+  disruption:
+    budgets:
+      - nodes: 0
+        schedule: "0 10 * * mon-fri"
+        duration: 16h
+        reasons:
+          - drifted
+          - expired
+      - nodes: 100%
+        reasons:
+          - empty
+      - nodes: "10%"
+      - nodes: 5
+    expireAfter: 720h | Never
+    consolidationPolicy: WhenUnderutilized | WhenEmpty
+    consolidateAfter: 1m | Never # Added to allow additional control over consolidation aggressiveness
+  weight: 10
+  limits:
+    cpu: "1000"
+    memory: 1000Gi
+status:
+  conditions:
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: NodeClassReady
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: Ready
+  resources:
+    cpu: "20"
+    memory: "8192Mi"
+    ephemeral-storage: "100Gi"
+```
+
+### Defaults
+
+Defaults are unchanged in the v1 API version but are included in the API proposal for completeness.
+
+`disruption.budgets` : `{"nodes": "10%"}`
+`disruption.expireAfter`: `30d/720h`
+`disruption.consolidationPolicy`: `WhenUnderutilized`
+`disruption.consolidateAfter`: `15s`
+
+### Printer Columns
+
+**Category:** Stability, Breaking
+
+#### Current
+
+```
+➜  karpenter git:(main) ✗ kubectl get nodepools -o wide
+NAME      NODECLASS   WEIGHT
+default   default     100
+fallback  fallback    
+```
+
+#### Proposed
+
+```
+➜  karpenter git:(main) ✗ kubectl get nodepools -o wide
+NAME      NODECLASS  NODES  READY  AGE   WEIGHT  CPU%  MEMORY%
+default   default    4      True   2d7h  100     5.8%  99%
+fallback  default    100    True   2d7h          20%   20%
+```
+
+**Standard Columns**
+
+1. Name
+2. NodeClass - Allows users to easily see the NodeClass that the NodePool is using
+3. Nodes - Allow users to easily view the number of nodes that are associated with a NodePool
+4. Ready - NodePools now have status conditions and will only be used for scheduling when ready. This readiness should be easily viewable by users
+5. Age - This is a common field for all Kubernetes objects and should be maintained in the standard set of columns
+
+**Wide Columns (-o wide)**
+
+1. Weight - Viewing the NodePools that will be evaluated first should be easily observable but may not be immediately useful to all users, particularly if the NodePools are named in a way that already indicate their ordering e.g. suffixed with fallback
+2. CPU% - The capacity of the CPU for all nodes provisioned by this NodePool as a percentage of its limits
```

Comment:
Would love to see pod utilization eventually

Reply:
Pod utilization / Total Capacity? Honestly, this strikes me as a more useful thing to show in printer columns than the capacity / limit; however, we have to be cautious around the number of writes that either of these updates would generate. Both the capacity values and the pod utilization values are going to change extremely quickly, so if we go with something like this we're going to have to concede that it can only be updated on a poll rather than on a watch

Reply:
pod utilization / spec.limit.pod if limit was set there. It would be cool to see a more generic approach where if a limit is set then we can show utilization of that resource limit

Reply:
We ended up choosing to drop this from the printer columns for now. Messing with printer columns was confirmed by api-machinery to not be a breaking change so we may add these percentages in later when CRDs support the ability to perform math calculations from data in the status.

For now, this change would require us to duplicate data into our status -- which feels unnecessary right now when this data can be captured from metrics.